File: Problem27_DiagonalTraverse.java
// Time Complexity : O(n), n -> total number of elements in the matrix
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Yes
// Three line explanation of solution in plain english: Starting at the top left corner, we have to iterate through the two dimensional array by keeping track of the direction that you are moving(up or down). Will handle the case when you reach the top or bottom or left and right borders.  

// Your code here along with comments explaining your approach

class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int height = mat.length - 1;
        int width = mat[0].length - 1;
        int[] result = new int[(height + 1) * (width + 1)];
        int row = 0, col = 0, idx = 0;
        boolean goingDown = true;
        while(!isOutOfBounds(row, col, height, width)) {
            result[idx++] = mat[row][col];
            if(!goingDown) {
                if(col == 0 || row == height) {
                    goingDown = true;
                    if(row == height) {
                        col++;
                    } else {
                        row++;
                    }
                } else {
                    row++;
                    col--;
                }
            } else {
                if(row == 0 || col == width) {
                    goingDown = false;
                    if(col == width) {
                        row++;
                    } else {
                        col++;
                    }
                } else {
                    row--;
                    col++;
                }
            }
        }
        return result;
    }
    public static boolean isOutOfBounds(int row, int col, int height, int width) {
        return row < 0 || row > height || col < 0 || col > width;
    }
}

File: Problem-27 Diagonal Traverse.py
# 498. Diagonal Traverse
# https://leetcode.com/problems/diagonal-traverse/

# Time complexiety: O(nm) 
# Space complexiety: O(1)

class Solution:
    def findDiagonalOrder(self, mat):
        direction = 0 # 0 - up, 1 - down
        res = []
        n = len(mat)
        m = len(mat[0])

        row = 0
        col = 0

        while len(res) < n*m:
            res.append(mat[row][col])
            if direction == 0:
                if row == 0:
                    if col == m-1:
                        row += 1
                        direction = 1
                    else:
                        col += 1
                        direction = 1
                elif col == m-1:
                    row += 1
                    direction = 1
                else:
                    row -= 1
                    col += 1
            else:
                if col == 0:
                    if row == n-1:
                        col += 1
                        direction = 0
                    else:
                        row += 1
                        direction = 0
                elif row == n-1:
                    col += 1
                    direction = 0
                else:
                    col -= 1
                    row += 1
        return res

obj = Solution()
obj.findDiagonalOrder([[1,2,3],[4,5,6],[7,8,9]])
obj.findDiagonalOrder([[1,2],[3,4]])
